src/Parser/TokenIterator.php

```diff
@@ -180,12 +184,30 @@ public function tryConsumeTokenValue(string $tokenValue): bool
 			return false;
 		}
 
-		$this->index++;
-		$this->skipIrrelevantTokens();
+		$this->next();
 
 		return true;
 	}
 
+	/**
+	 * @return Comment[]
+	 */
+	public function flushComments(): array
```

Comment:
i think this PR is starting to be in a pretty good shape. I still have to go through the many test cases and think about it very hard (and try to break your logic :))

Every time there's a new stateful logic introduced like here with the comments array, I try to think how we can prevent it being used wrongly.

I think TokenIterator should throw an exception if the comments weren't flushed (the array is not empty):

1) Definitely when consuming TOKEN_END
2) Maybe even more often, basically every time we assume the array should be empty. Like when consuming some other token than EOL and maybe others.

Reply:
I'm having a look through the code, too, but so far I don't actually see any calls to `consumeTokenType` with `TOKEN_END` as an argument.

My next thought was to do a check before returning from `TypeParser::parse`, but that one is recursive, so `TypeParser` doesn't really know when it's done (without some further refactoring).

`PhpDocParser::parse` has some potential, but while it is possible to have comments remaining when the parsing is done, it's not clear to me if it's an error. For example, this seems like it should be valid to me:

```php

/**
 *  @return int|false
 *
 * // can return false on failure
 * /

```

Since we always attach each comment to the following node, and there are no more nodes after the comment, we technically have a leftover comment, but no cause to throw an exception.

Reply:
Alright, so this comment made me hear a loud sound of breaking tableware :D

We need to make sure that comments outside of types, outside of TypeParser, are still parsed as they are today!

So for example:

```php
/** @param int $a // this is a description */
```

Must become a description of ParamTagValueNode. Please be aware that there are two differents implementations of consuming descriptions: `parseOptionalDescription` and `parseOptionalDescriptionAfterDoctrineTag`. So make sure to update and test both. Thank you.

In this case:

```php
/**
 *  @return int|false
 *
 * // can return false on failure
 * /
```

This must become a standalone PhpDocTextNode as today.

Can you make sure they still do?

Reply:
> Can you make sure they still do?

Did they ever? I didn't know about `parseOptionalDescription` or `parseOptionalDescriptionAfterDoctrineTag`, but I checked out the `1.23.x` branch, stuck a breakpoint in each of them, and ran the test. Neither breakpoint was hit.

Can you provide a breaking test for me? Feel free to push a commit.

Reply:
> /** @param int $a // this is a description */
> Must become a description of ParamTagValueNode

By the way, I'm not sure about that one. None of the samples I see in the tests look like that. They all look like this:

```php
/** 
@param int $a this is a description 
*/
```

Reply:
Here it is: https://github.com/phpstan/phpdoc-parser/commit/0bb2fe4cc43a6144d3135facec4d036258a5e4dd

Please squash and rebase your branch so that the history is linear and clean. I was unable to do so.

Reply:
Wonderful, thank you. I've rebased everything down to just our two commits so we can get a better look at things.

As much as I hate to disagree with you, I feel I must question your premise. I feel that you are getting comments and descriptions mixed-up. A description is text that comes after a tag value, and a comment denotes text as having no functional meaning. So, a commented description is _not_ a description, it's simply a comment. The same goes for a comment sitting on a line by itself.

I've spent many hours debugging nikic parser, and what I've learned is that comments are always attached to the following token, and never the preceding one.

So, in order to achieve what you are asking I would need to violate both the fundamental functionality of the parser this one is modeled after, and also the principle of least surprise, in that we are stripping the "do nothing" nature of comments.

That said, I still don't know how nikic parser handles comments that come at the end of a PHP file and have no tokens to attach to, but if you will bear with me I will find out.

Reply:
> how nikic parser handles comments that come at the end of a PHP file

The answer seems to be: it does nothing with them.

They are parsed, a new `Comment` instance is created, and it is added to the running `startAttributes` object kept around during the parse algorithm, but it is not ultimately added to any nodes.

Reply:
Maybe this will help illustrate how I'm viewing the difference. Compare:

```php
/**
 * @param int $foo the number of foos to count
 * /
 ```

vs

```php
/**
 * @param int $foo // TODO: come up with description
 * /
 ```

Reply:
@ondrejmirtes please let me know your thoughts.

Reply:
> Otherwise we're already inside a comment. There's no structural difference between @param int $foo the number of foos to count and @param int $foo // TODO: come up with description.

There is a functional difference. Let's rewrite it a different way:

```php
// with description
$tag = [
  'name' => '@param',
  'type' => `int`,
  'parameterName' => '$foo,
 'description` => 'the number of foos to count'
];
```

```php
// with comment
$tag = [
  'name' => '@param',
  'type' => `int`,
  'parameterName' => '$foo,
 'description` => '' // TODO: come up with a description
];

```

Is that more clear? The first one has a description, the second one doesn't.

> // outside of PHPDoc types must mean the same thing as before. If not for nothing else then at least for backward compatibility.

Can you elaborate on this? I'm not clear on what you feel is broken. Comments are currently ignored, which is what is supposed to happen. How exactly have I broken backwards-compatibility? All the existing tests pass, and comments are neatly ignored. 

If you want to add the kind of complexity you have in mind later that's your business, but for now do we really have a problem?!

Reply:
> How exactly have I broken backwards-compatibility?

Ah, maybe I understand. You're saying that there is code currently floating around out in the wild where there are already comments assigned to descriptions and being interpreted as descriptions, and your concern is that after this change they will lose their description status, is that right?

Reply:
Yes, we can't change the AST like that. It'd also make altering these comments more difficult and diferent for the users of format-preserving printer, most notably Slevomat CS.

Reply:
Okay. At this point I am going to abandon this PR. I tried. I feel that the behavior in this PR is correct. I understand it may "break" descriptions out in the wild, but I would argue that the code out in the wild is incorrect, and this fixes it. That happens sometimes in the software development cycle, and the solution is to do a major version release, not add tons of complexity trying to achieve backwards complexity with existing behavior that is wrong.

I'm going to use my own fork, which I really hate to do, but I've done everything I can within reason to solve this problem rationally, and I'm not getting support. Good luck.